src/liballoc/arc.rs

```diff
@@ -282,22 +296,29 @@ pub fn strong_count<T: ?Sized>(this: &Arc<T>) -> usize { Arc::strong_count(this)
 /// let _y = x.clone();
 /// assert!(get_mut(&mut x).is_none());
 /// # }
-/// # }
 /// ```
 #[inline]
 #[unstable(feature = "arc_unique")]
-#[deprecated(since = "1.2.0",
-             reason = "this function is unsafe with weak pointers")]
-pub unsafe fn get_mut<T: ?Sized>(this: &mut Arc<T>) -> Option<&mut T> {
-    // FIXME(#24880) potential race with upgraded weak pointers here
-    if Arc::strong_count(this) == 1 && Arc::weak_count(this) == 0 {
-        // This unsafety is ok because we're guaranteed that the pointer
-        // returned is the *only* pointer that will ever be returned to T. Our
-        // reference count is guaranteed to be 1 at this point, and we required
-        // the Arc itself to be `mut`, so we're returning the only possible
-        // reference to the inner data.
-        let inner = &mut **this._ptr;
-        Some(&mut inner.data)
+pub fn get_mut<T: ?Sized>(this: &mut Arc<T>) -> Option<&mut T> {
+    // lock the weak pointer count if we appear to be the sole weak pointer
+    // holder.
+    if this.inner().weak.compare_and_swap(1, usize::MAX, SeqCst) != 1 {
```

Comment:
Shouldn’t this read `this.inner().weak.compare_and_swap(1, usize::MAX, SeqCst) == 1`?